compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml

```diff
@@ -0,0 +1,49 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | MOVE Prelude interface functionality
+module DA.Internal.Interface (
+  HasInterfaceTypeRep(..),
+  HasToInterface(..),
+  HasFromInterface(..),
+  Implements,
+  toInterface,
+  toInterfaceContractId,
+  fromInterfaceContractId,
+) where
+
+import DA.Internal.Prelude
+import DA.Internal.Template.Functions
+import DA.Internal.LF
+
+class HasInterfaceTypeRep i where
+  interfaceTypeRep : i -> TypeRep
+
+class HasToInterface t i where
+  _toInterface : t -> i
+
+toInterface : forall i t. HasToInterface t i => t -> i
+toInterface = _toInterface
+
+class HasFromInterface t i where
+  fromInterface : i -> Optional t
+
+type Implements t i =
+  ( HasInterfaceTypeRep i
+  , HasToInterface t i
+  , HasFromInterface t i
+  )
+
+toInterfaceContractId : forall i t. HasToInterface t i => ContractId t -> ContractId i
+toInterfaceContractId = coerceContractId
+
+fromInterfaceContractId : forall t i. (HasFromInterface t i, HasFetch i) => ContractId i -> Update (Optional (ContractId t))
+fromInterfaceContractId cid = do
+  iface <- fetch cid
+  pure $ case fromInterface iface of
+    None -> None
+    Some (_ : t) -> Some (coerceContractId cid)
```

Comment:
looks like `fmap`?

Reply:
Kind of, but the result doesn't really depend on the contents of the `Optional`. We _do_ want that `t` type signature there, though, and I think the explicit pattern match looks better than `fmap (\(_:t) -> ...)`

Reply:
`fmap @t`? if that doesn’t work no issue keeping it.

Reply:
ah fair, yeah `fmap @Some @t` would work. But that made me notice I could move the `@t` to the call to `fromInterface`and then `fmap`'s little sibling `$>` does the right thing.

Reply:
it's probably a bit painful to import `(DA.Functor.$>)` just for this, so I went with 
```diff
- pure $ case fromInterface iface of
-   None -> None
-   Some (_ : t) -> Some (coerceContractId cid)
+ pure $
+   const (coerceContractId cid) <$> fromInterface @t iface
```

Reply:
[bikeshed]
another way is to use `do` syntax: 
```
pure do
  fromInterface @t iface 
  Some (coerceContractId cid)
```
[/bikeshed]